### ISSUE TITLE
[Bug]: The accuracy of multiple cards and single card is inconsistent

### ISSUE THREAD

**Issue by jiangix-paper** [vllm-project/vllm#13801]
### Your current environment

vllm 0.7.1
torch 2.5.1
transformers 4.48.3

### 🐛 Describe the bug

I find the accuracy is inconsistent when I use tensor-parallel-size=2 compare with tensor-parallel-size=1
the commands are as follows:

CUDA_VISIBLE_DEVICES=5 python -m vllm.entrypoints.openai.api_server --model ./ --port 7863 --tensor-parallel-size=1

CUDA_VISIBLE_DEVICES=5,6 python -m vllm.entrypoints.openai.api_server --model ./ --port 7863 --tensor-parallel-size=2

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jiangix-paper** [comment]
@DarkLight1337 Hello, Can you help me?

**Comment by DarkLight1337** [comment]
Which model are you using?

**Comment by jiangix-paper** [comment]
Hello, I use the qwen2.5-coder-14B-Instruct model, and I use it to generate code or sql. And I turn off the sampling use the temperature = 0.

**Comment by jiangix-paper** [comment]
The model I used is finetuned based on qwen2.5-coder-14B-Instruct. In NL2SQL Scenario，the accuracy of multiple cards and single card is inconsistent. 
By the way, are the accuracy of single-card and multi-card completely equivalent? Or is there a deviation?

**Comment by DarkLight1337** [comment]
@youkaichao might know more.

**Comment by jiangix-paper** [comment]
@youkaichao Hello, can you help me please?

**Comment by tjtanaa** [comment]
@youkaichao @jiangix-paper @DarkLight1337 

To provide more data points I have attached my tests on A100 with vLLM v0.7.3 installed from pypi

## The script I am using for running BF16 across TP=1,2,4,8:

```python
from vllm import LLM, SamplingParams
import json
import torch
import numpy as np
import random

# Set the seed for PyTorch
torch.manual_seed(0)

# Set the seed for NumPy
np.random.seed(0)

# Set the seed for the built-in random module
random.seed(0)

TP=8
FILENAME_OUTPUT=f"meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp{TP}_32tokens_seed0.txt"

if __name__ == '__main__':
    sampling_params = SamplingParams(temperature=0, top_k=1, logprobs=10, max_tokens=32)
    llm = LLM("meta-llama/Meta-Llama-3-8B-Instruct", tensor_parallel_size=TP)
    response = llm.generate("What is batch inference?", sampling_params)

    with open(FILENAME_OUTPUT, "w") as f:
        f.write(str(response))
```
### The generated output files:

[meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp2_32tokens_seed0.txt](https://github.com/user-attachments/files/19244261/meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp2_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp4_32tokens_seed0.txt](https://github.com/user-attachments/files/19244260/meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp4_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp8_32tokens_seed0.txt](https://github.com/user-attachments/files/19244262/meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp8_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp1_32tokens_seed0.txt](https://github.com/user-attachments/files/19244259/meta-llama_Meta-Llama-3-8B-Instruct_bf16_bf16kvcache_tp1_32tokens_seed0.txt)

### Summary of the sampled output:
***TP1:*** text=' Inference is the process of using a trained machine learning model to make predictions or classify new, unseen data. Batch inference is a type of inference where the model', token_ids=(763, 2251, 374, 279, 1920, 315, 1701, 264, 16572, 5780, 6975, 1646, 311, 1304, 20492, 477, 49229, 502, 11, 64233, 828, 13, 35004, 45478, 374, 264, 955, 315, 45478, 1405, 279, 1646)

_**TP2:**_ text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new, unseen batch of', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 11, 64233, 7309, 315)

***TP4:*** text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)

***TP8:*** text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)


## The script I am using for running BF16 across TP=1,2,4,8:

```python
from vllm import LLM, SamplingParams
import json
import torch
import numpy as np
import random

# Set the seed for PyTorch
torch.manual_seed(0)

# Set the seed for NumPy
np.random.seed(0)

# Set the seed for the built-in random module
random.seed(0)

TP=8
FILENAME_OUTPUT=f"meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp{TP}_32tokens_seed0.txt"

if __name__ == '__main__':
    sampling_params = SamplingParams(temperature=0, top_k=1, logprobs=10, max_tokens=32)
    llm = LLM("meta-llama/Meta-Llama-3-8B-Instruct", tensor_parallel_size=TP, quantization='fp8', kv_cache_dtype='fp8')
    response = llm.generate("What is batch inference?", sampling_params)

    with open(FILENAME_OUTPUT, "w") as f:
        f.write(str(response))
```
### The generated output files:

[meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp1_32tokens.txt](https://github.com/user-attachments/files/19244303/meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp1_32tokens.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp2_32tokens.txt](https://github.com/user-attachments/files/19244301/meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp2_32tokens.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp4_32tokens.txt](https://github.com/user-attachments/files/19244302/meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp4_32tokens.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp8_32tokens.txt](https://github.com/user-attachments/files/19244304/meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp8_32tokens.txt)


### Summary of the sampled output:
***TP1:*** text=' Inference is the process of using a trained machine learning model to make predictions or classify new, unseen data. Batch inference refers to the process of using a trained', token_ids=(763, 2251, 374, 279, 1920, 315, 1701, 264, 16572, 5780, 6975, 1646, 311, 1304, 20492, 477, 49229, 502, 11, 64233, 828, 13, 35004, 45478, 19813, 311, 279, 1920, 315, 1701, 264, 16572)

***TP2:*** text=' Inference is the process of using a trained machine learning model to make predictions or classify new, unseen data. Batch inference is a type of inference where the model', token_ids=(763, 2251, 374, 279, 1920, 315, 1701, 264, 16572, 5780, 6975, 1646, 311, 1304, 20492, 477, 49229, 502, 11, 64233, 828, 13, 35004, 45478, 374, 264, 955, 315, 45478, 1405, 279, 1646)

***TP4:*** text=' Inference is the process of using a trained machine learning model to make predictions on new, unseen data. Batch inference is a type of inference where the model is', token_ids=(763, 2251, 374, 279, 1920, 315, 1701, 264, 16572, 5780, 6975, 1646, 311, 1304, 20492, 389, 502, 11, 64233, 828, 13, 35004, 45478, 374, 264, 955, 315, 45478, 1405, 279, 1646, 374)

***TP8:*** text=' Inference is the process of using a trained machine learning model to make predictions or classify new, unseen data. Batch inference is a type of inference where the model', token_ids=(763, 2251, 374, 279, 1920, 315, 1701, 264, 16572, 5780, 6975, 1646, 311, 1304, 20492, 477, 49229, 502, 11, 64233, 828, 13, 35004, 45478, 374, 264, 955, 315, 45478, 1405, 279, 1646)

**Comment by tjtanaa** [comment]
@jiangix-paper I have new leads.

Llama model weight by default is stored as `bfloat16`, if we force vLLM to convert the weight to `half` (`float16`), the generated token is more deterministics.

### FP16 script

```
from vllm import LLM, SamplingParams
import json
import torch
import numpy as np
import random

# Set the seed for PyTorch
torch.manual_seed(0)

# Set the seed for NumPy
np.random.seed(0)

# Set the seed for the built-in random module
random.seed(0)

TP=8
FILENAME_OUTPUT=f"meta-llama_Meta-Llama-3-8B-Instruct_fp8quant_fp8kvcache_tp{TP}_32tokens_seed0.txt"

if __name__ == '__main__':
    sampling_params = SamplingParams(temperature=0, top_k=1, logprobs=10, max_tokens=32)
    llm = LLM("meta-llama/Meta-Llama-3-8B-Instruct", tensor_parallel_size=TP, dtype='half')
    response = llm.generate("What is batch inference?", sampling_params)

    with open(FILENAME_OUTPUT, "w") as f:
        f.write(str(response))
```

### Summary of the sampled output:
TP1: text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)

TP2: text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)

TP4: text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)

TP8: text=' Batch inference is a technique used in machine learning where a model is trained on a batch of data and then used to make predictions on a new batch of data.', token_ids=(35004, 45478, 374, 264, 15105, 1511, 304, 5780, 6975, 1405, 264, 1646, 374, 16572, 389, 264, 7309, 315, 828, 323, 1243, 1511, 311, 1304, 20492, 389, 264, 502, 7309, 315, 828, 13)


### The generated output files:

[meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp1_32tokens_seed0.txt](https://github.com/user-attachments/files/19250000/meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp1_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp2_32tokens_seed0.txt](https://github.com/user-attachments/files/19249997/meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp2_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp4_32tokens_seed0.txt](https://github.com/user-attachments/files/19249999/meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp4_32tokens_seed0.txt)
[meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp8_32tokens_seed0.txt](https://github.com/user-attachments/files/19249998/meta-llama_Meta-Llama-3-8B-Instruct_fp16_fp16kvcache_tp8_32tokens_seed0.txt)

### Discussion 
Even if the sampled token is the same, if we inspect the logprobs value for the first generated token `Batch` , all four TPs have different logprob values

TP1: 35004: Logprob(logprob=-1.3351420164108276, rank=1, decoded_token=' Batch')
TP2: 35004: Logprob(logprob=-1.3422571420669556, rank=1, decoded_token=' Batch')
TP4: 35004: Logprob(logprob=-1.3448426723480225, rank=1, decoded_token=' Batch')
TP8: 35004: Logprob(logprob=-1.3457964658737183, rank=1, decoded_token=' Batch')

**Comment by jiangix-paper** [comment]
Yeah, you are right. So I think the reason is different precision between bf16 and fp16
